das/dastree/dastree.go

```diff
@@ -0,0 +1,210 @@
+	//
+	//      0   1   2       <=>  leaf n = H(0xfe, H(bin n))                    step 0
+	//
+	//  Where H is keccak and L is the length
+	//  Intermediate hashes like '*' from above may be recorded via the `record` closure
+	//
+
+	keccord := func(value []byte) bytes32 {
+		hash := crypto.Keccak256Hash(value)
+		record(hash, value)
+		return hash
+	}
+	prepend := func(before byte, slice []byte) []byte {
+		return append([]byte{before}, slice...)
+	}
+
+	unrolled := arbmath.ConcatByteSlices(preimage...)
+	if len(unrolled) == 0 {
+		return keccord(prepend(LeafByte, keccord([]byte{}).Bytes()))
+	}
+
+	length := uint32(len(unrolled))
+	leaves := []node{}
+	for bin := uint32(0); bin < length; bin += BinSize {
+		end := arbmath.MinUint32(bin+BinSize, length)
+		hash := keccord(prepend(LeafByte, keccord(unrolled[bin:end]).Bytes()))
+		leaves = append(leaves, node{hash, end - bin})
+	}
+
+	layer := leaves
+	for len(layer) > 1 {
+		prior := len(layer)
+		after := prior/2 + prior%2
+		paired := make([]node, after)
+		for i := 0; i < prior-1; i += 2 {
+			firstHash := layer[i].hash.Bytes()
+			otherHash := layer[i+1].hash.Bytes()
+			sizeUnder := layer[i].size + layer[i+1].size
+			dataUnder := firstHash
+			dataUnder = append(dataUnder, otherHash...)
+			dataUnder = append(dataUnder, arbmath.Uint32ToBytes(sizeUnder)...)
+			parent := node{
+				keccord(prepend(NodeByte, dataUnder)),
+				sizeUnder,
+			}
+			paired[i/2] = parent
+		}
+		if prior%2 == 1 {
+			paired[after-1] = layer[prior-1]
+		}
+		layer = paired
+	}
+	return layer[0].hash
+}
+
+func Hash(preimage ...[]byte) bytes32 {
+	// Merkelizes without recording anything. All but the replay binary's DAS will call this
+	return RecordHash(func(bytes32, []byte) {}, preimage...)
+}
+
+func HashBytes(preimage ...[]byte) []byte {
+	return Hash(preimage...).Bytes()
+}
+
+func FlatHashToTreeHash(flat bytes32) bytes32 {
+	// Forms a degenerate dastree that's just a single leaf
+	// note: the inner preimage may be larger than the 64 kB standard
+	return crypto.Keccak256Hash(append([]byte{LeafByte}, flat[:]...))
+}
+
+func ValidHash(hash bytes32, preimage []byte) bool {
+	if hash == Hash(preimage) {
+		return true
+	}
+	if len(preimage) > 0 {
+		kind := preimage[0]
+		return kind != NodeByte && kind != LeafByte && hash == crypto.Keccak256Hash(preimage)
+	}
+	return false
+}
+
+func Content(root bytes32, oracle func(bytes32) []byte) ([]byte, error) {
+	// Reverses hashes to reveal the full preimage under the root using the preimage oracle.
+	// This function also checks that the size-data is consistent and that the hash is canonical.
+	//
+	// Notes
+	//     1. Because we accept degenerate dastrees, we can't check that single-leaf trees are canonical.
+	//     2. For any canonical dastree, there exists a degenerate single-leaf equivalent that we accept.
+	//     3. Only the committee can produce trees unwrapped by this function
```

Comment:
Maybe note that only the replay binary calls this and that it accepts old-style flat hashes

Reply:
I've updated the comment